Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -315,311 +545,498 @@ void CGContextSetFillColorSpace(CGContextRef pContext, CGColorSpaceRef colorSpac
+}
+
+static void __CGContextDrawGeometry(CGContextRef context, ID2D1Geometry* geometry, CGPathDrawingMode drawMode) {
+    ComPtr<ID2D1RenderTarget> renderTarget = context->RenderTarget();
+    auto& currentState = context->CurrentGState();
+
+    ComPtr<ID2D1DeviceContext> deviceContext;
+    FAIL_FAST_IF_FAILED(renderTarget.As(&deviceContext));
+
+    D2D1_SIZE_F targetSize = renderTarget->GetSize();
+    ComPtr<ID2D1Image> originalTarget;
+    deviceContext->GetTarget(&originalTarget);
+
+    deviceContext->BeginDraw();
+
+    ComPtr<ID2D1CommandList> commandList;
+    FAIL_FAST_IF_FAILED(deviceContext->CreateCommandList(&commandList));
+    deviceContext->SetTarget(commandList.Get());
+
+    CGAffineTransform& currentTransform = currentState.transform;
+    CGAffineTransform transform = CGAffineTransformScale(currentTransform, 1.0, -1.0);
+    transform = CGAffineTransformTranslate(transform, 0., targetSize.height);
+    deviceContext->SetTransform({ transform.a, transform.b, transform.c, transform.d, transform.tx, transform.ty });
+
+    if (drawMode & kCGPathFill) {
+        if (drawMode & kCGPathEOFill) {
+            // TODO(DH): GH#1077 Regenerate geometry in Even/Odd fill mode.
+        }
+        deviceContext->FillGeometry(geometry, currentState.fillBrush.Get());
+    }
+
+    if (drawMode & kCGPathStroke) {
+        ComPtr<ID2D1Factory> factory;
+        deviceContext->GetFactory(&factory);
+
+        // TODO(DH): GH#1077 Do not recreate for every drawing operation
+        ComPtr<ID2D1StrokeStyle> strokeStyle;
+        // dashes must be adjusted to be based on line width
+        std::vector<CGFloat> adjustedDashes{ currentState.dashes };
+        for (float& f : adjustedDashes) {
+            f /= currentState.lineWidth;
+        }
+        FAIL_FAST_IF_FAILED(
+            factory->CreateStrokeStyle(currentState.strokeProperties, adjustedDashes.data(), adjustedDashes.size(), &strokeStyle));
+
+        deviceContext->DrawGeometry(geometry, currentState.strokeBrush.Get(), currentState.lineWidth, strokeStyle.Get());
+    }
+
+    deviceContext->EndDraw();
+    commandList->Close();
+
+    deviceContext->BeginDraw();
+    deviceContext->SetTarget(originalTarget.Get());
+
+    bool layer = false;
+    if (currentState.alpha != 1.0f || false /* mask, clip, etc. */) {
+        layer = true;
+        renderTarget->PushLayer(D2D1::LayerParameters(D2D1::InfiniteRect(),
+                                                      NULL,
+                                                      D2D1_ANTIALIAS_MODE_PER_PRIMITIVE,
+                                                      D2D1::IdentityMatrix(),
+                                                      currentState.alpha),
+                                nullptr);
     }
 
-    ctx->Backing()->CGContextClearRect(rct);
+    deviceContext->DrawImage(commandList.Get());
+
+    if (layer) {
+        renderTarget->PopLayer();
+    }
+
+    FAIL_FAST_IF_FAILED(deviceContext->EndDraw());
+}
```

Comment:
There are instances in which the draw target will need to be recreated, but we have perhaps lost the recipe.